### ISSUE TITLE
Route.Hosts on KongIngress does not alter resulting Route.Hosts

### ISSUE THREAD

**Issue by llinvillegwre** [Kong/kubernetes-ingress-controller#1164]
### Summary

Kong-Ingress-Controller provides a convenient means of setting multiple SNI values but does not provide a convenient means of setting multiple Hosts. When an organization has thousands of _Ingress_ objects (each with a dozen or more routes) there's no convenient means of matching multiple Host values to the same set of routes. While doubling _Ingress_ objects or _Ingress_ rules is a possibility, such a strategy does not scale well for organizations that already have a large number of _Ingress_ objects and routes.

Setting _route.hosts_ on _KongIngress_ object does not override (or append) to hosts on route(s)

- **Suggestion 1**: Include KongIngress.Route.Hosts in overrideByKongIngress logic
- **Suggestion 2 (Recommended)**: Include new Ingress annotation **konghq.com/host-aliases** that accepts a comma delimited list of hosts. Annotation should result in additional Hosts being appended to Route Hosts. Annotation should not allow duplicates. Annotation should enrich Route Hosts for all Routes parsed from the Ingress definition.

**Kong Ingress controller version**
1.1.1

**Kong or Kong Enterprise version**
2.3.2.0 Enterprise DB-less

### Kubernetes version

1.17

### What happened
Setting _route.hosts_ on _KongIngress_ object is ignored.

### Expected behavior
Setting _route.hosts_ on _KongIngress_ object should result in _KongIngress.Route.Hosts_ values being appended to route(s). Alternatively, an _Ingress_ annotation (e.g. _konghq.com/hosts_) would be useful.

### Steps To Reproduce
Just try to deploy a _KongIngress_ or _Ingress_ with multiple host values without defining multiple _Ingress_ rules.

```
apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  name: example
  namespace: example
proxy: {}
route:
  https_redirect_status_code: 308
  # hosts is ignored here
  hosts:
  - exampletwo.com
  - examplethree.com
  protocols:
  - https
  strip_path: false
```

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    konghq.com/override: example
    # Possible to specify multiple SNIs but not possible to specify multiple HOSTs without a lot of duplication in Kong config
    konghq.com/snis: exampleone.com,exampletwo.com,examplethree.com
    # Should a new annotation be added? i.e. konghq.com/hosts: exampleone.com,exampletwo.com,examplethree.com
    kubernetes.io/ingress.class: kong
  name: example
  namespace: example
spec:
  rules:
  - host: exampleone.com
    http:
      paths:
      - backend:
          serviceName: example
          servicePort: 80
        path: /rest/account/
      - backend:
          serviceName: example
          servicePort: 80
        path: /rest/customer/
      - backend:
          serviceName: example
          servicePort: 80
        path: /rest/order/
```

**Comment by llinvillegwre** [comment]
In regards to **Suggestion 1**. I was thinking something like the following in **route.go**:

```
// overrideByKongIngress sets Route fields by KongIngress
func (r *Route) overrideByKongIngress(log logrus.FieldLogger, kongIngress *configurationv1.KongIngress) {
	if kongIngress == nil || kongIngress.Route == nil {
		return
	}
	ir := kongIngress.Route
	...
	if len(ir.Hosts) != 0 {
		r.Hosts = append(r.Hosts, ir.Hosts...)
	}
	...
}
```

**Comment by mflendrich** [comment]
Hi @llinvillegwre,

This is indeed a limitation of the Kubernetes Ingress API. It will be solved in Gateway API; [`HTTPRoute`'s `Hostname` field](https://gateway-api.sigs.k8s.io/spec/#networking.x-k8s.io/v1alpha1.Hostname) accepts a list of hostnames.

For the time being (until KIC supports Gateway API's `HTTPRoute`), it would indeed be possible and IMO reasonable (@hbagdi any comments?) to introduce a workaround like the one you proposed. But we (maintainers) will rather invest in support for Gateway API, than provide this feature in `KongIngress`. Your PR contributing a `route.hosts` field support for KongIngress would be welcome, though.

**Comment by hbagdi** [comment]
KongIngress resource *only* allows to set fields that are not already present in Kubernetes manifests and this is very intentional because we do not want multiple sources of any piece of data. Features like these make debugging pretty hard for most users.

This problem is very real but I'd like us to figure a solution in which all the "hosts" info is present in the Ingress resource itself. So, an annotation on the Ingress resource could be a good way forward. Have other Ingress Controllers figured out a solution to this problem that we can adopt as well?

**Comment by llinvillegwre** [comment]
I'm willing to submit a PR. Is **Suggestion 2**, creating a new annotation on the _Ingress_ object the desired solution?

**Comment by hbagdi** [comment]
Can you either update the existing description or provide a more detailed summary of the design? Please cover the definition of the annotation, how it interacts with existing hosts in the spec and give some examples and list cover corner cases and solutions for those corner cases.

**Comment by llinvillegwre** [comment]
OK. I will provide an update shortly.

**Comment by rainest** [comment]
Aside from my general reservations around having both a resource-standard (`Ingress.spec.rules[].host`) and vendor-specific (annotation) way for populating a value (the Kong route hostnames), I believe the proposed solution in #1272 poses an issue. Consider this Ingress:

```
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: example
  annotations:
    konghq.com/host-aliases: "foo.example,bar.example"
spec:
  rules:
  - host: example.com
    http:
      paths:
      - path: /foo
        backend:
          serviceName: service1
  - host: example.net
    http:
      paths:
      - path: /foo
        backend:
          serviceName: service2
```
Since the aliases apply Ingress-wide, each rule gets the same override and they'll generate overlapping routes, `{"hosts": "example.com,foo.example,bar.example", "paths":"/foo"}` and `{"hosts": "example.net,foo.example,bar.example", "paths":"/foo"}`, and a request for `foo.example/foo` would have ambiguous matching.

Such overlapping routes are permitted, so while this is a footgun, the resulting config is at least valid. I don't think there's any way to avoid this problem in design or code: it requires user vigilance to either only use this on Ingresses with one standard host or to ensure that no rules share paths.

@llinvillegwre do you see any way that could avoid that?

@hbagdi @mflendrich do you think the utility of working around the Ingress limitation outweighs the usability concern?

**Comment by llinvillegwre** [comment]
@rainest I had the same concern / question (e.g. annotation applying across all rules).

The annotation is really intended to be used in cases where one is not defining multiple hosts via multiple rules; therefore, one solution might be ignoring the annotation when the spec contains multiple rules (or perhaps when rules reference more than one host). However, I can see where having an annotation that's conditionally applied based on the number of rules or hosts may become a source of confusion. 

Therefore, I kept the solution simple (apply the annotation consistently across all rules) until I got more opinions / feedback. If the annotation needs to be conditional I can update the PR as necessary.
